servlet/src/main/java/io/grpc/servlet/ServletAdapter.java

```diff
@@ -0,0 +1,327 @@
+      logger.log(FINEST, "[{0}] headers: {1}", new Object[] {logId, headers});
+    }
+
+    Long timeoutNanos = headers.get(TIMEOUT_KEY);
+    if (timeoutNanos == null) {
+      timeoutNanos = 0L;
+    }
+    asyncCtx.setTimeout(TimeUnit.NANOSECONDS.toMillis(timeoutNanos));
+    StatsTraceContext statsTraceCtx =
+        StatsTraceContext.newServerContext(streamTracerFactories, method, headers);
+
+    ServletServerStream stream = new ServletServerStream(
+        asyncCtx,
+        statsTraceCtx,
+        maxInboundMessageSize,
+        attributes.toBuilder()
+            .set(
+                Grpc.TRANSPORT_ATTR_REMOTE_ADDR,
+                new InetSocketAddress(req.getRemoteHost(), req.getRemotePort()))
+            .set(
+                Grpc.TRANSPORT_ATTR_LOCAL_ADDR,
+                new InetSocketAddress(req.getLocalAddr(), req.getLocalPort()))
+            .build(),
+        getAuthority(req),
+        logId);
+
+    transportListener.streamCreated(stream, method, headers);
+    stream.transportState().runOnTransportThread(stream.transportState()::onStreamAllocated);
+
+    asyncCtx.getRequest().getInputStream()
+        .setReadListener(new GrpcReadListener(stream, asyncCtx, logId));
+    asyncCtx.addListener(new GrpcAsycListener(stream, logId));
+  }
+
+  private static Metadata getHeaders(HttpServletRequest req) {
+    Enumeration<String> headerNames = req.getHeaderNames();
+    checkNotNull(
+        headerNames, "Servlet container does not allow HttpServletRequest.getHeaderNames()");
+    List<byte[]> byteArrays = new ArrayList<>();
+    while (headerNames.hasMoreElements()) {
+      String headerName = headerNames.nextElement();
+      Enumeration<String> values = req.getHeaders(headerName);
+      if (values == null) {
+        continue;
+      }
+      while (values.hasMoreElements()) {
+        String value = values.nextElement();
+        if (headerName.endsWith(Metadata.BINARY_HEADER_SUFFIX)) {
+          byteArrays.add(headerName.getBytes(StandardCharsets.US_ASCII));
+          byteArrays.add(BaseEncoding.base64().decode(value));
+        } else {
+          byteArrays.add(headerName.getBytes(StandardCharsets.US_ASCII));
+          byteArrays.add(value.getBytes(StandardCharsets.US_ASCII));
+        }
+      }
+    }
+    return InternalMetadata.newMetadata(byteArrays.toArray(new byte[][]{}));
+  }
+
+  private static String getAuthority(HttpServletRequest req) {
+    try {
+      return new URI(req.getRequestURL().toString()).getAuthority();
+    } catch (URISyntaxException e) {
+      logger.log(FINE, "Error getting authority from the request URL {0}" + req.getRequestURL());
+      return req.getServerName() + ":" + req.getServerPort();
+    }
+  }
+
+  /**
+   * Call this method when the adapter is no longer needed. The gRPC server will be terminated.
+   */
+  public void destroy() {
+    transportListener.transportTerminated();
+  }
+
+  private static final class GrpcAsycListener implements AsyncListener {
```

Comment:
typo? Asyc -> Async

Reply:
Yeah. Thanks. I also noticed that just had not a chance to fix yet.